### PR TITLE
fix payment discount (iOS Promotional Offers) into being actually usable

### PR DESCRIPTION
**Pull request by crystalstorm**
Make iOS Promotional Offers in in_app_purchase plugin usable

Resolves: [[in-app-purchase][iOS] Promotional Offers in iOS not working](https://github.com/flutter/flutter/issues/107533#issue-1303299545)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart [hunk 1]
```diff
@@ -81,6 +81,20 @@ class InAppPurchaseStoreKitPlatform extends InAppPurchasePlatform {
     return true; // There's no error feedback from iOS here to return.
   }
 
+  static Future<bool> buyNonConsumableWithDiscount(
+      {required PurchaseParam purchaseParam, SKPaymentDiscountWrapper? discount}) async {
+    await _skPaymentQueueWrapper.addPayment(SKPaymentWrapper(
```
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:86]: Why is it a static method while it is using an instance variable?

Comment (cyanglaz) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:86]: I would imagine the discount can be part of the `PurchaseParam`

**Review by cyanglaz — CHANGES_REQUESTED** [review]
Also needs tests

**Comment by stuartmorgan** [comment]
Thanks for the contribution! You’ve checked boxes in the PR checklist above that are not reflected in this PR, so I’m assuming this is a work in progress and am marking it as a Draft. Please review the checklist, updating the PR as appropriate, and when the checklist reflects the state of the PR as posted please feel free to mark it as ready for review.

Also, it would be helpful if you could provide more information in the issue about what exactly you were doing and what didn't work; it just asserts that the feature doesn't work without much detail, which makes it hard to evaluate.

**Comment by crystalstorm** [comment]
@stuartmorgan indeed it's work in progress.

About missing details, will this be helpful enough:

> I was trying to implement iOS Promotional offers using the in-app-purchase plug in. I did not found any example of how it could be done in the original repo or articles in the web. It seems like nobody has done it, or mentioned how it is doable. I found the Pull request where the Promotional offers were implemented in the plugin: [here](https://github.com/flutter/plugins/pull/4458) . But there was no example of using it either. Browsing trough the code I found out there is no direct way to use newly added `SKPaymentDiscountWrapper` class upon which the feature relies. There are 2 other issues related with the current implementation mentioned in my comment [here](https://github.com/flutter/flutter/issues/107533#issuecomment-1183310299)

and if so, do you need it posted in the original issue body text?

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart (hunk 1 above)
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:86]: @cyanglaz the instance variable is also static. I agree it is not the best option. I merely wanted a working solution to open a pull request with. Then have a discussion about how to resolve it, or have it resolved by someone from the maintainers.

If it's part of PurchaseParam, this method would not be needed at all but I presume there should be also Android counterpart . In the original Pull request I saw only iOS implementation. I don't know how the Android offers work (if there are any).

**Comment by stuartmorgan** [comment]
Since this is marked as a draft and hasn't been updated in several months I'm going to close it to clean out our review queue. Please don't hesitate to submit a new PR if you decide to revisit this. Thanks!

**Comment by crystalstorm** [comment]
@stuartmorgan I've opened another [PR here](https://github.com/flutter/plugins/pull/6541) which isn't a draft. Hopefully it will be more useful than this one.

**Pull request closed without merging**